sdk/filter/filter.go

```diff
@@ -0,0 +1,44 @@
+package filter
+
+import "github.com/hypertrace/goagent/sdk"
+
+// Filter runs on request evaluates whether request should be
+// blocked
+type Filter struct {
```

Comment:
how are all these filters going to be installed?  I was thinking we could mirror https://github.com/hypertrace/javaagent/blob/main/filter-api/src/main/java/org/hypertrace/agent/filter/MultiFilter.java so we'd have something like:

`
type MultiFilter {
  filters []filter
}

NewMultiFilter(filters []filter) *MultiFilter {
...
}


type filter interface {
  evaluateRequestURL        func(span sdk.Span, URL string) bool
  evaluateRequestHeaders func(span sdk.Span, headers map[string][]string) bool
  evaluateRequestBody       func(span sdk.Span, body []byte) bool
}
`

Reply:
It will look like what's in `handler_test.go`. I agree this is nicer, but I'll keep filter as struct so that we can easily tell if the filter cares about body or not, and can easily avoid body extraction if it doesn't and body capture is not configured. (@jcchavezs's idea).

Reply:
I think that may be overkill. For the cases where there's a filter configured, we can assume we're always going to have body capture enabled.

Even if the particular filter isn't making use of the body, there's no harm in it having the implement `evaluateRequestBody` but have it a noop.

Reply:
I'm ok with tying it with body data capture.

Reply:
I really like consistency but I also believe we should stick to contracts and not implementations. I think every language has its own trade-offs and implementation strategies. As of Go I think the right approach is using a struct and we can keep consistency on the contract, we will probably do the same in languages that don't have interfaces (e.g. javascript, Ruby). Personally (as I told @davexroth) I'd prefer instrumentation to have the less impact on the user code and reading the body for nothing does not sound the best approach in Go (maybe in other languages where you don't need to reassign the body every time you read it). On the other hand I think we have spent too much time on this back and forth so I'd let you guys decide.

Reply:
I think I'm missing the case where we're reading the body un-necessarily (which is bad!). can you expand on that case?

Reply:
Users don't necessarily read the body, e.g. unsupported content-type (there are other cases but let's assume they all happen in an upper layer, e.g. authorization). In our case we will anyways read the body on behalf of the user and the user will probably dismiss the request without making use of the body content. By doing that we are impacting the user workflow and it can be more intensive if we talk about big payloads.